### PR TITLE
Fix install windows features when install state is removed

### PR DESCRIPTION
**Pull request by dheerajd-msys**
Signed-off-by: dheerajd-msys <dheeraj.dubey@msystechnologies.com>

windows features were not getting installed when it's InstallState was `Removed`
## Description

- This change handles installing features when state is removed or disabled(available)
- Works for single as well as multi-feature installation
- Earlier intersection of the features returned nil when state was `Removed` so feature was not getting installed.
- Ensured `Idempotent`  for the `windows_feature_powershell`
- Redundant test removed
## Related Issue

Fixes https://github.com/chef/chef/issues/9536
## O/P Postfix
```
PS C:\Users\Administrator\chef> Get-WindowsFeature -Name NET-Framework-Core

Display Name                                            Name                       Install State
------------                                            ----                       -------------
    [ ] .NET Framework 3.5 (includes .NET 2.0 and 3.0)  NET-Framework-Core               Removed


PS C:\Users\Administrator\chef> bundle exec chef-apply \\tsclient\Downloads\recipe-windows-feature.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_feature[NET-Framework-Core] action install
    * windows_feature_powershell[NET-Framework-Core] action install
      - install Windows feature net-framework-core

PS C:\Users\Administrator\chef> Get-WindowsFeature -Name NET-Framework-Core

Display Name                                            Name                       Install State
------------                                            ----                       -------------
    [X] .NET Framework 3.5 (includes .NET 2.0 and 3.0)  NET-Framework-Core             Installed


PS C:\Users\Administrator\chef>
PS C:\Users\Administrator\chef> Get-WindowsFeature -Name Web-Net-Ext45

Display Name                                            Name                       Install State
------------                                            ----                       -------------
            [ ] .NET Extensibility 4.6                  Web-Net-Ext45                  Available


PS C:\Users\Administrator\chef> Get-WindowsFeature -Name Web-Net-Ext45

Display Name                                            Name                       Install State
------------                                            ----                       -------------
            [ ] .NET Extensibility 4.6                  Web-Net-Ext45                  Available


PS C:\Users\Administrator\chef> bundle exec chef-apply C:\Users\Administrator\Downloads\recipe-windows-feature.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_feature[Web-Asp-Net45, Web-Net-Ext45] action install
    * windows_feature_powershell[Web-Asp-Net45, Web-Net-Ext45] action install
      - install Windows features web-asp-net45,web-net-ext45

PS C:\Users\Administrator\chef> Get-WindowsFeature -Name Web-Net-Ext45

Display Name                                            Name                       Install State
------------                                            ----                       -------------
            [X] .NET Extensibility 4.6                  Web-Net-Ext45                  Installed


```
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/resource/windows_feature_powershell.rb [hunk 1]
```diff
@@ -141,8 +141,8 @@ def to_formatted_array(x)
       action_class do
         # @return [Array] features the user has requested to install which need installation
         def features_to_install
-          # the intersection of the features to install & disabled features are what needs installing
-          @install ||= new_resource.feature_name & node["powershell_features_cache"]["disabled"]
+          # the intersection of the features to install & disabled/removed features are what needs installing
+          @install ||= new_resource.feature_name & ( node["powershell_features_cache"]["disabled"] | node["powershell_features_cache"]["removed"] )
```
Comment (tas50) [inline lib/chef/resource/windows_feature_powershell.rb:145]: This won't work for any user that hasn't provided a source. We need to add a union of removed features against the existing list of features that can be installed when a source is available.

Comment (dheerajd-msys) [reply lib/chef/resource/windows_feature_powershell.rb]: Yes @tas50 Verified and updated for same. It is working as expected now for single as well as multi-feature installation

File: spec/unit/resource/windows_feature_powershell_spec.rb [hunk 2]
```diff
@@ -35,10 +35,6 @@
     expect(resource.feature_name).to eql(%w{snmp dhcp})
   end
 
-  it "sets the default action as :install" do
-    expect(resource.action).to eql([:install])
-  end
```
Comment (lamont-granquist) [inline spec/unit/resource/windows_feature_powershell_spec.rb:40]: this shouldn't change?

Comment (dheerajd-msys) [reply spec/unit/resource/windows_feature_powershell_spec.rb:40]: @lamont This test was redundant so removed one of them https://github.com/chef/chef/blob/master/spec/unit/resource/windows_feature_powershell_spec.rb#L30:L38

File: lib/chef/resource/windows_feature_powershell.rb [hunk 3]
```diff
@@ -141,8 +141,12 @@ def to_formatted_array(x)
...
         # @return [Array] features the user has requested to install which need installation
         def features_to_install
-          # the intersection of the features to install & disabled features are what needs installing
-          @install ||= new_resource.feature_name & node["powershell_features_cache"]["disabled"]
+          # the intersection of the features to install & disabled/removed features are what needs installing
+          if new_resource.source && node["powershell_features_cache"]["removed"]
+            @install ||= new_resource.feature_name & ( node["powershell_features_cache"]["disabled"] | node["powershell_features_cache"]["removed"] )
+          else
+            @install ||= new_resource.feature_name & node["powershell_features_cache"]["disabled"]
+          end
```
Comment (phiggins) [inline lib/chef/resource/windows_feature_powershell.rb:149]: * From reading `reload_cached_powershell_data`, both of these values will always be present and will always be an array, so doing a union is safe in bot cases.
* This code looks ok, but in general memoizing a value inside a conditional can introduce subtle caching bugs. I'd invert that to make it more clear what's happening here.

```suggestion
          # the intersection of the features to install & disabled/removed features are what needs installing
          @features_to_install || = begin
            features = node["powershell_features_cache"]["disabled"]
            features |= node["powershell_features_cache"]["removed"] if new_resource.source
         
            new_resource.feature_name & features
          end
```

Comment (phiggins) [reply lib/chef/resource/windows_feature_powershell.rb:149]: Some unit tests for the different cases would be a good addition.

Comment (dheerajd-msys) [reply lib/chef/resource/windows_feature_powershell.rb]: Thanks @phiggins updated per suggestion!

File: spec/unit/resource/windows_feature_powershell_spec.rb [hunk 4]
```diff
@@ -54,4 +50,14 @@
...
+
+  it "install a single feature" do
+    resource.feature_name "snmp"
+    expect { resource.action :install }.not_to raise_error
+  end
+
+  it "install multi feature" do
+    resource.feature_name "SNMP, DHCP"
+    expect { resource.action :install }.not_to raise_error
+  end
```
Comment (phiggins) [inline spec/unit/resource/windows_feature_powershell_spec.rb:62]: These tests don't test the behavior that was added in this PR.

What do you think about some tests like:

```
it "does not attempt to install features that have been removed" do
  node["powershell_features_cache"]["disabled"] = ["dhcp"]
  node["powershell_features_cache"]["removed"] = ["snmp"]
  resource.feature_name "dhcp, "snmp"

  expect(resource.features_to_install).to eq(["dhcp"])
end

it "attempts to install features that have been removed when source is set" do
  node["powershell_features_cache"]["disabled"] = ["dhcp"]
  node["powershell_features_cache"]["removed"] = ["snmp"]
  resource.feature_name "dhcp, "snmp"
  resource.source "some appropriate value"

  expect(resource.features_to_install).to eq(["dhcp", "snmp"])
end
```

Comment (dheerajd-msys) [reply spec/unit/resource/windows_feature_powershell_spec.rb:62]: In process to add these tests, Was getting `undefined method `[]=' for nil:NilClass` while implementing these.

Comment (phiggins) [reply spec/unit/resource/windows_feature_powershell_spec.rb:62]: Depending on what the error was, you might need to add something like `node["powershell_features_cache"] ||= {}`.

Comment (dheerajd-msys) [reply spec/unit/resource/windows_feature_powershell_spec.rb:63]: Resolved now and updated, Thank you @phiggins !!
